Scripts/BuildDatabank/AddData.py

```diff
@@ -16,572 +35,550 @@
+                            df_files,
+                            pd.DataFrame(
+                                [
+                                    {
+                                        "NAME": file_provided[0],
+                                        "TYPE": key_sim,
+                                        "REQUIRED": is_required,
+                                        "SIZE_MB": file_size_mb,
+                                        "HASH": file_hash,
+                                    }
+                                ]
+                            ),
+                        ],
+                        ignore_index=True,
+                    )
+                    files_list.append([file_provided[0], file_hash])
+
+                    # Find the keys of the required files to calculate the master_hash
+                    if is_required:
+                        sha1_list_requied.append(file_hash)
+
+                    sim_hashes[key_sim] = files_list  # TODO Problematic
+        except KeyError as e:  # It is notmal that fails for "ID" and "SOFTWARE"
+            continue
+
+    logger.info(f"Summary of downloaded files:{os.linesep}")
+    print(df_files)
+    print()
+
+    # Calculate the hash of a file contaning the hashes of each of the required files
+    # This should be always invariant as it will be used unique identifier for a simualtion
+    # Note order the hashes of the required files before calculating the hash (That means that the required files cannot change)
+
+    # Calculates numbers of lipid molecules in each leaflet. This is done by checking on which side of the centre
+    # of mass the membrane each the centre of mass of a lipid molecule is.
+    # If a lipid molecule is split so that headgroup and tails are their own residues, the centre of mass of the
+    # headgroup is used in the calculation.
+    ################################################################################################################
 
-## Check that the number of atoms between data and README.yaml match
+    logger.info(
+        "Calculating the numbers of lipid molecules in each leaflet based on the center of mass of the membrane and lipids."
+    )
+    logger.info(
+        "If a lipid molecule is split to multiple residues, the centre of mass of the headgroup is used."
+    )
 
-number_of_atomsTRJ = len(u.atoms)
+    top = ""
+    traj = ""
 
-number_of_atoms = 0
-for key_mol in all_molecules:
-    mapping_dict = {}
-    try:
-        mapping_file = os.path.join(
-            os.getcwd(), "mapping_files", sim["COMPOSITION"][key_mol]["MAPPING"]
-        )
-    except:
-        continue
+    if sim["SOFTWARE"] == "gromacs":
+        top = os.path.join(dir_tmp, sim["TPR"][0][0])
+        traj = os.path.join(dir_tmp, sim["TRJ"][0][0])
+    elif sim["SOFTWARE"] == "openMM" or sim["SOFTWARE"] == "NAMD":
+        traj = os.path.join(dir_tmp, sim["TRJ"][0][0])
+        top = os.path.join(dir_tmp, sim["PDB"][0][0])
```

Comment:
Someone should write test case for openMM and NAMD data and he should not be me. Otherwise I would consider that they don't work. And they probably don't even before this PR. I think NAMD analysis I did tested somewhere.

Reply:
These things has already been done before me. 

The dictionary was defined here: [databank_defs.py:888](https://github.com/NMRLipids/Databank/blob/main/Scripts/BuildDatabank/databank_defs.py#L888-L895)

And it's used from [AddData.py](https://github.com/NMRLipids/Databank/blob/main/Scripts/BuildDatabank/AddData.py#L107-L114):
```
# validate yaml entries and return updated sim dict
    try:
        sim, files = parse_valid_config_settings(info_yaml)
    except KeyError as e:
```

But it doesn't allow to get rid of these messy if-s completely. Because there exists TOP-file for one engine, and it can be provided. But not all the functions read this TOP-file. It's done very bad all over the code. It's not only here where you meet it. 

If we want to put thing on the proper track then we need to agree on how topology files from different engines should be used. Should we just require them? Do we allow using structure files like PDB/GRO instead? Do we check on the molecule integrity inside? This is the problem mainly. Now it's a complete mess which have already created improper analysis files (I'm pretty sure about it).

Reply:
I think, work on issue #205 will cover it. The core was done before, but the developers ignored that by some reasons. Maybe the reasons did exist, for example there were problems with *MDanalysis* and PSF.

Reply:
What is missing in `engine_dict` is the `REQUIRED_FIELDS` entry. For any engine it can be organized as follows:

```python
"REQUIRED_FIELDS": {
    "TOP": (names for top (in GROMACS case "TPR"),
    "TRJ": (names for traj),
    # other required fields.
}
```

What can be done next, is a function that generates topology and trajectory variables:
```python
def top_traj_from_sim(sim) -> tuple[str, str]:
    top = software_dict[sim["SOFTWARE"]]["REQUIRED_FIELDS"]["TOP"][0]  #if there are multiple options for top, then a slightly more complicated logics should be used
    trj = software_dict[sim["SOFTWARE"]]["REQUIRED_FIELDS"]["TRJ"][0]  #if there are multiple options for trj, then a slightly more complicated logics should be used
    return top, trj

try:
     top, traj = top_traj_from_sim(sim)
except Exception:
    logger.error(
     "SOFTWARE '%s' is not a proper option.\n"
     "Use either 'gromacs', 'openMM', or 'NAMD'.")
    quit()
```

This can be a cleaner way, in my opinion.

Reply:
I did. Look commit 8ca880bc.